file/file_prefetch_buffer.cc

```diff
@@ -315,17 +315,41 @@ Status FilePrefetchBuffer::PrefetchAsyncInternal(
 
     // Move data from curr_ buffer to third.
     CopyDataToBuffer(curr_, offset, length);
+    // swap the buffers.
+    curr_ = curr_ ^ 1;
+    second = curr_ ^ 1;
+
+    // All the bytes are in curr_. second buffer can go for async prefetching
+    // before copying data since second is empty.
+    if (offset + length <=
+        bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize()) {
+      uint64_t rounddown_start =
+          bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize();
+      uint64_t roundup_end =
+          Roundup(rounddown_start + readahead_size, alignment);
+      uint64_t roundup_len = roundup_end - rounddown_start;
+      uint64_t chunk_len = 0;
+      CalculateOffsetAndLen(alignment, rounddown_start, roundup_len, second,
+                            false /*refit_tail*/, chunk_len);
+
+      // Update the buffer offset.
+      bufs_[second].offset_ = rounddown_start;
+      assert(roundup_len >= chunk_len);
+      uint64_t read_len2 = static_cast<size_t>(roundup_len - chunk_len);
+      ReadAsync(opts, reader, read_len2, chunk_len, rounddown_start, second)
+          .PermitUncheckedError();
+    }
+
+    CopyDataToBuffer(curr_, offset, length);
+
+    // Length == 0: All the requested data has been copied to third buffer and
+    // it has gone for only async prefetching so it can return without doing
+    // anything further. Length > 0: More data needs to be consumed so it will
+    // continue async and sync prefetching and copy the remaining data to third
+    // buffer in the end.
     if (length == 0) {
```

Comment:
Don't need this check because in the main check we already make sure data is overlapping with `offset + length > bufs_[second].offset_`